File: backend/app/outbox.py
"""
Transactional outbox writer.

`emit()` INSERTs a durable event row on the SAME connection/transaction as the
domain change that produced it. Because get_scoped_connection() wraps each
request in a single transaction, the domain write and its outbox event either
both commit or both roll back — there is no window where the meeting is
completed but the event was lost, or vice-versa.

A separate worker process (app/workers/outbox_worker.py) drains the table after
commit and fans out to summaries / notifications / calendar sync. Realtime
Broadcast, when adopted, is published from that worker — never from the client.
"""
import json


async def emit(conn, event_type: str, *, aggregate_id: str, tenant_id: str | None = None,
               payload: dict | None = None, event_version: int = 1,
               aggregate_type: str = "meeting") -> str:
    """Append one event to the outbox in the caller's transaction. Returns event_id."""
    row = await conn.fetchrow(
        """
        INSERT INTO meeting_outbox (event_type, event_version, aggregate_type, aggregate_id, tenant_id, payload)
        VALUES ($1, $2, $3, $4, $5, $6::jsonb)
        RETURNING event_id
        """,
        event_type, event_version, aggregate_type, aggregate_id,
        (tenant_id if tenant_id else None), json.dumps(payload or {}),
    )
    return str(row["event_id"])

File: apps/backend-node/src/common/outbox.ts
import { ScopedSql } from '../database/database.service';

/**
 * Transactional outbox writer. Faithful port of backend/app/outbox.py.
 *
 * `emit()` INSERTs a durable event row on the SAME transaction as the domain
 * change that produced it. Because DatabaseService.scoped() wraps each request
 * in a single transaction, the domain write and its outbox event either both
 * commit or both roll back — no window where the meeting is completed but the
 * event was lost. A separate worker (apps/backend-node-worker) drains
 * meeting_outbox after commit and fans out.
 */
export async function emit(
  sql: ScopedSql,
  eventType: string,
  opts: {
    aggregateId: string;
    tenantId?: string | null;
    payload?: Record<string, unknown> | null;
    eventVersion?: number;
    aggregateType?: string;
  },
): Promise<string> {
  const rows = await sql`
    INSERT INTO meeting_outbox (event_type, event_version, aggregate_type, aggregate_id, tenant_id, payload)
    VALUES (
      ${eventType},
      ${opts.eventVersion ?? 1},
      ${opts.aggregateType ?? 'meeting'},
      ${opts.aggregateId},
      ${opts.tenantId ?? null},
      ${sql.json((opts.payload ?? {}) as never)}
    )
    RETURNING event_id
  `;
  return String(rows[0].event_id);
}
